pymodbus/repl/server/main.py

```diff
@@ -178,29 +178,29 @@ def run(
     data_block_settings = modbus_config.pop("data_block_settings", {})
     modbus_config = modbus_config.get(modbus_server, {})
     modbus_config = process_extra_args(extra_args, modbus_config)
-    if modbus_server != "serial":
-        handler = modbus_config.pop("handler", "ModbusConnectedRequestHandler")
-    else:
-        handler = modbus_config.pop("handler", "ModbusSingleRequestHandler")
-    handler = DEFUALT_HANDLERS.get(handler.strip())
 
-    modbus_config["handler"] = handler
     modbus_config["randomize"] = randomize
     modbus_config["change_rate"] = change_rate
-    app = ReactiveServer.factory(
-        modbus_server,
-        framer,
-        modbus_port=modbus_port,
-        slave=modbus_slave_id,
-        loop=loop,
-        single=False,
-        data_block_settings=data_block_settings,
-        **web_app_config,
-        **modbus_config,
-    )
-    loop.run_until_complete(app.run_async(repl))
+
+    async def _wrapper():
+        app = ReactiveServer.factory(
+            modbus_server,
+            framer,
+            modbus_port=modbus_port,
+            slave=modbus_slave_id,
+            single=False,
+            data_block_settings=data_block_settings,
+            **web_app_config,
+            **modbus_config,
+        )
+        await app.run_async(repl)
+        return app
+
+    app = loop.run_until_complete(_wrapper())
```

Comment:
This seems wrong, at least you are repeating it within the "if repl".

Reply:
As far as I understand, the wrapper function (and so app.run_async()) does not actually start the full server. It only does the initialization work (which, nevertheless, must be started in a running loop).

The _wrapper() just initialises the app (the same code for repl and non-repl case) and starts the web_app but does not run the modbus server itself.

The original [code](https://github.com/pymodbus-dev/pymodbus/blob/08529612b11966e4d16d085b7550d74737472b67/pymodbus/repl/server/main.py#L201C1-L205C27) did exactly the same 2-step process (first create the objects + start web server and as second step run the whole app):

```
    loop.run_until_complete(app.run_async(repl))


    if repl:
        loop.run_until_complete(run_repl(app))
    else:
        loop.run_forever()
```

The issue of the original code was, that the Server objects (and the start of the web app) must be initialized in a running loop in some cases. Otherwise, there would be a problem like this e.g. in the UDP case:

```
pymodbus.server run -s udp
#Traceback (most recent call last):

  File "/usr/bin/pymodbus.server", line 8, in <module>
    sys.exit(app())
             ^^^^^

  File "/usr/lib/python3.11/site-packages/pymodbus/repl/server/main.py", line 183, in run
    app = ReactiveServer.factory(
          ^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/lib/python3.11/site-packages/pymodbus/server/reactive/main.py", line 499, in factory
    server = server(
             ^^^^^^^

  File "/usr/lib/python3.11/site-packages/pymodbus/server/async_io.py", line 473, in __init__
    self.loop = asyncio.get_running_loop()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^

RuntimeError: no running event loop
```

Reply:
sound just about right.